### PR TITLE
Flight properties

### PR DESCRIPTION
**Pull request by kylep-dremio**
Add optional arbitrary properties to flight calls.

### CODE REVIEW

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightConstants.java
Comment (jduo) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightConstants.java]: Header names are usually dash-separated.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/PropertyCallOption.java
Comment (jduo) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/PropertyCallOption.java]: What if the client did not send any properties?

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerPropertyMiddleware.java
Comment (jduo) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerPropertyMiddleware.java]: This needs to be indented to 4 spaces past the p in public. Checkstyle won't complain about this.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerPropertyMiddleware.java
Comment (jduo) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerPropertyMiddleware.java]: Throw on malformed properties?

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerPropertyMiddleware.java
Comment (jduo) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerPropertyMiddleware.java]: Private constructor?

File: java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestCallOptions.java
Comment (jduo) [inline java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestCallOptions.java]: Let's add more tests with multiple properties, properties with special chars in name/value.

**Review by jduo — CHANGES_REQUESTED** [review]

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/PropertyCallOption.java
Comment (kylep-dremio) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/PropertyCallOption.java]: Value should be a serializable.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerPropertyMiddleware.java
Comment (kylep-dremio) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerPropertyMiddleware.java]: Also updated ServerCallHeaderAuthMiddleware as that's where this was taken from.

**Review by jduo — APPROVED** [review]
LGTM. Let's not merge this to flight-auth-redesign, then once that gets merged to apache/master, make the PR against that.

**Pull request closed without merging**
